shared/chat/conversation/messages/wrapper/exploding-height-retainer/index.native.js

```diff
@@ -34,53 +69,164 @@ class ExplodingHeightRetainer extends React.Component<Props, State> {
           !!this.state.height && this.props.retainHeight && {height: this.state.height},
         ])}
       >
-        {this.props.retainHeight && <AshTower numImages={this.state.numImages} />}
-        {this.props.retainHeight &&
-          (!this.props.explodedBy ? (
-            <Text type="BodySmall" style={styles.exploded}>
-              EXPLODED
-            </Text>
-          ) : (
-            <Text lineClamp={1} type="BodySmall" style={styles.exploded}>
-              EXPLODED BY{' '}
-              <ConnectedUsernames
-                type="BodySmallSemibold"
-                clickable={true}
-                usernames={[this.props.explodedBy]}
-                inline={true}
-                colorFollowing={true}
-              />
-            </Text>
-          ))}
-        {!this.props.retainHeight && this.props.children}
+        {this.state.children}
+        <AnimatedAshTower
+          exploded={this.props.retainHeight}
+          explodedBy={this.props.explodedBy}
+          numImages={this.state.numImages}
+        />
       </Box>
     )
   }
 }
 
-const AshTower = (props: {numImages: number}) => {
+type AshTowerProps = {exploded: boolean, explodedBy: ?string, numImages: number}
+type AshTowerState = {width: NativeAnimated.Value}
+class AnimatedAshTower extends React.Component<AshTowerProps, AshTowerState> {
+  state = {width: this.props.exploded ? new NativeAnimated.Value(100) : new NativeAnimated.Value(0)}
+
+  componentDidUpdate(prevProps: AshTowerProps) {
+    if (!prevProps.exploded && this.props.exploded) {
+      // just exploded! animate
+      NativeAnimated.timing(this.state.width, {
+        duration: animationDurationMs,
+        easing: NativeEasing.inOut(NativeEasing.ease),
+        toValue: 100,
+      }).start()
+    }
+  }
+
+  render() {
+    const width = this.state.width.interpolate({
+      inputRange: [0, 100],
+      outputRange: ['0%', '100%'],
+    })
+    return (
+      <NativeAnimated.View style={[{width}, styles.slider]}>
+        <AshTower {...this.props} />
+        <EmojiTower animatedValue={this.state.width} numImages={this.props.numImages} />
+      </NativeAnimated.View>
+    )
+  }
+}
+
+class EmojiTower extends React.Component<
+  {numImages: number, animatedValue: NativeAnimated.Value},
+  {running: boolean}
+> {
+  state = {running: false}
+  componentDidMount() {
+    this.props.animatedValue.addListener(this._listener)
```

Comment:
docs don't mention this at all but removeListener should exist and we should call it

Reply:
it's buried but here it is: https://facebook.github.io/react-native/docs/animatedvalue.html
I added a `removeAllListeners`